modepy/spaces.py

```diff
@@ -64,6 +80,71 @@ def space_for_shape(shape: Shape, order: int) -> FunctionSpace:
 # }}}
 
 
+# # {{{ generic tensor product space
+
+class TensorProductSpace(FunctionSpace):
+    """
+    .. attribute:: bases
+
+        A :class:`tuple` of the base spaces that take part in the
+        tensor product.
+
+    .. automethod:: __init__
+    """
+
+    bases: Tuple[FunctionSpace, ...]
+
+    # NOTE: https://github.com/python/mypy/issues/1020
+    def __new__(cls, bases: Tuple[FunctionSpace, ...]) -> Any:
+        if len(bases) == 1:
+            return bases[0]
+        else:
+            return FunctionSpace.__new__(cls)
+
+    def __init__(self, bases: Tuple[FunctionSpace, ...]) -> None:
+        self.bases = sum([
+            space.bases if isinstance(space, TensorProductSpace) else (space,)
+            for space in bases
+            ], ())
+
+    @property
+    def spatial_dim(self) -> int:
+        return sum(space.spatial_dim for space in self.bases)
+
+    @property
+    def space_dim(self) -> int:
+        return np.prod([space.space_dim for space in self.bases])
+
+    def __repr__(self) -> str:
+        return (f"{type(self).__name__}("
+                f"spatial_dim={self.spatial_dim}, space_dim={self.space_dim}, "
+                f"bases={self.bases!r}"
+                ")")
+
+
+@space_for_shape.register(TensorProductShape)
+def _space_for_tensor_product_shape(
+        shape: TensorProductShape,
+        order: Union[int, Tuple[int, ...]]) -> TensorProductSpace:
```

Comment:
Added a bit of documentation to 60e35cb.

Now that you mention it, I'm not convinced it's a good idea to allow this in the first place. The reason I added it was mostly so that the tests could construct different spaces through `space_for_shape` (otherwise they would need to special case tensor products).

I'm fine with removing it if you think it's cleaner.

Reply:
> How come?

From what I remember, mostly because the tuple version is specific to the tensor product and it felt weird to add that to the general interface.

It's a fairly small/self-contained/documented change though, so I don't feel particularly strongly either way.